jvm/src/common/Language/Java/Unsafe.hs

```diff
@@ -319,7 +327,7 @@ $(mkVariadic [t| J ('Class $(TH.varT (TH.mkName "sym"))) |] $ \ctx typ pats args
 -- @
 --
 -- You can pass any number of 'Coercible' arguments to the constructor.
-new :: New r => r
+new :: Variadic "new" r => r
```

Comment:
A challenge with variadic programming is that Haskell considers functions to be just like all other data. But a variadic function can't return a function, or else we don't know how many parameters there are. That's why having a result in `IO` works easily, but have a general monad does not work as easily. Still, I agree that this could be generalized.

Reply:
`Variadic` itself isn't exported, so the documentation for it won't help much. What we have is an explanation of the `Variadic` constraint at the module level, and in the documentation of each function.

A standalone version of `Variadic` would be very nice. But it would have to look different than it does now. Indexing with a symbol allows us to use the exact same name as the function: `Variadic "foo"` says `foo` is variadic. But if `Variadic` were to escape the scope of the current module, then we'd have to be careful that no other module exports a variadic function also called `foo`. The fix is to use an empty datatype instead.

I'm okay with generalizing `Variadic` with the extra indices that you suggest, especially if `Variadic` is to be spun out, though I'd keep `r :: *` as in your first example. Because it will be hard to have an `r :: Constraint -> * -> *` be anything other than a data/newtype.

Keep in mind that even with extra indices, `Variadic` is still a special case of variadic functions. The prototypical one is `printf`, where one would normally want the freedom to specify arbitrary serializers/formats for each argument, rather than be constrained to canonical serializers given by type class instances (like the [non-type-class-based one](https://tidsskrift.dk/brics/article/download/19284/16911) does). The inline-java case is a special case.